include/sys/metaslab_impl.h

```diff
@@ -129,6 +130,74 @@ typedef enum trace_alloc_type {
 #define	WEIGHT_GET_COUNT(weight)		BF64_GET((weight), 0, 55)
 #define	WEIGHT_SET_COUNT(weight, x)		BF64_SET((weight), 0, 55, x)
 
+#define	CAS_MAGIC		0xb10ca110ca7e57a7ULL
+#define	CAS_HISTOGRAM_SIZE	9
+
+/*
+ * Each alloction stat is comprised
+ * of a byte count and block count.
+ */
+struct alloc_stat {
+	uint64_t as_bytes;
+	uint64_t as_blocks;
+};
+
+/*
+ * Persistent per-class block allocation stats
+ *
+ * Note since this object is stored as a generitc DMU_OTN_UINT64_METADATA
+ * type, we include a magic value (blockallocatestat)
+ */
+typedef struct class_alloc_stats_phys {
+	uint64_t	cas_magic;
+	uint64_t	cas_birth_txg;		/* start of stat collection */
+
+	struct alloc_stat cas_dedup_data;	/* dedup specific data */
+	struct alloc_stat cas_dmu_dnodes;	/* dmu dnode block */
+	struct alloc_stat cas_file_data;	/* file leaf blocks */
+	struct alloc_stat cas_file_indirect;	/* file metadata */
+	struct alloc_stat cas_zvol_data;	/* zvol leaf blocks */
+	struct alloc_stat cas_zvol_indirect;	/* zvol metadata */
+
+	/*
+	 * These histograms track allocated blocks by size.  Each
+	 * bucket contains the number of allocated blocks whose
+	 * size is >= to a power-of-two and < the next power-of two
+	 * for the range of 2^9 through 2^17.
+	 */
+	uint64_t	cas_metadata_histogram[CAS_HISTOGRAM_SIZE];
+	uint64_t	cas_smallblk_histogram[CAS_HISTOGRAM_SIZE];
+} class_alloc_stats_phys_t;
+
+/*
+ * Block context required to categorize block types in allocation stats
+ */
+struct block_context {
```

Comment:
I was a bit confused because I thought this was the "context" required to determine which class the block should be in. Maybe rename this something like `block_stats_type_t` or `class_stats_target_t`, so that it's more closely associated with the stats.  It's essentially telling us what "type" this is for stats purposes, i.e. what buckets we should be incrementing/decrementing.

Reply:
We have switched to block accounting (instead of per dva) so this context is no longer needed and has been removed.